client/consul.go

```diff
@@ -0,0 +1,159 @@
+
+var _ ConsulClientInterface = (*ConsulClient)(nil)
+
+// ConsulClientInterface is an interface for a Consul Client
+// If more consul client functionality is required, this interface should be extended with the following
+// considerations:
+// Each request to Consul is:
+// - Retried
+// - Logged at DEBUG-level
+// - Easily mocked
+type ConsulClientInterface interface {
+	GetLicense(ctx context.Context, q *consulapi.QueryOptions) (string, error)
+	GetSKU(ctx context.Context) (string, error)
+}
+
+// ConsulClient is a client to the Consul API
+type ConsulClient struct {
+	*consulapi.Client
+	retry  retry.Retry
+	logger logging.Logger
+}
+
+// Self represents the response body from Consul /v1/agent/self API endpoint.
+// Care must always be taken to do type checks when casting, as structure could
+// potentially change over time.
+type Self = map[string]map[string]interface{}
+
+// NewConsulClient constructs a consul api client
+func NewConsulClient(conf *config.Config, maxRetry int) (*ConsulClient, error) {
+	consulConf := conf.Consul
+	transport := hcat.TransportInput{
+		SSLEnabled: *consulConf.TLS.Enabled,
+		SSLVerify:  *consulConf.TLS.Verify,
+		SSLCert:    *consulConf.TLS.Cert,
+		SSLKey:     *consulConf.TLS.Key,
+		SSLCACert:  *consulConf.TLS.CACert,
+		SSLCAPath:  *consulConf.TLS.CAPath,
+		ServerName: *consulConf.TLS.ServerName,
+
+		DialKeepAlive:       *consulConf.Transport.DialKeepAlive,
+		DialTimeout:         *consulConf.Transport.DialTimeout,
+		DisableKeepAlives:   *consulConf.Transport.DisableKeepAlives,
+		IdleConnTimeout:     *consulConf.Transport.IdleConnTimeout,
+		MaxIdleConns:        *consulConf.Transport.MaxIdleConns,
+		MaxIdleConnsPerHost: *consulConf.Transport.MaxIdleConnsPerHost,
+		TLSHandshakeTimeout: *consulConf.Transport.TLSHandshakeTimeout,
+	}
+
+	consul := hcat.ConsulInput{
+		Address:      *consulConf.Address,
+		Token:        *consulConf.Token,
+		AuthEnabled:  *consulConf.Auth.Enabled,
+		AuthUsername: *consulConf.Auth.Username,
+		AuthPassword: *consulConf.Auth.Password,
+		Transport:    transport,
+	}
+
+	clients := hcat.NewClientSet()
+	if err := clients.AddConsul(consul); err != nil {
+		return nil, err
+	}
+
+	logger := logging.Global().Named(loggingSystemName).Named(consulSubsystemName)
+
+	r := retry.NewRetry(maxRetry, time.Now().UnixNano())
+	return &ConsulClient{
+		Client: clients.Consul(),
+		retry:  r,
+		logger: logger}, nil
+}
+
+// GetLicense queries Consul for a signed license, and returns it if available
+func (c *ConsulClient) GetLicense(ctx context.Context, q *consulapi.QueryOptions) (string, error) {
+	c.logger.Debug("getting license")
+
+	desc := "consul client get license"
+	var err error
+	var license string
+
+	err = c.retry.Do(ctx, func(context.Context) error {
+		license, err = c.Operator().LicenseGetSigned(q)
+		if err != nil {
+			license = ""
+			return err
+		}
+		return nil
+	}, desc)
+
+	return license, err
+}
+
+// GetSKU queries Consul for information about itself, it then
```

Comment:
Interesting, thanks for sharing. I'm not familiar with "SKU" in an engineering context - I might be the only one though? What are your thoughts on a name that doesn't require prior knowledge? Maybe something like AgentInfo? In particular I think it would be helpful to have something more descriptive for users in the error message and logging

Reply:
agentInfo isn't right either though, since we aren't getting all the info, we just want to know if it's Enterprise or OSS.

Version doesn't really work either, since we don't care about the number... I know we use "repo" in our release guide, but I don't think that really makes sense here. I can't think of anything other than SKU here.